ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdRecalculationService.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2015 Avanza Bank AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.avanza.ymer;
+
+import static com.avanza.ymer.MirroredObject.DOCUMENT_INSTANCE_ID;
+import static com.avanza.ymer.MirroredObject.DOCUMENT_ROUTING_KEY;
+import static com.avanza.ymer.util.GigaSpacesInstanceIdUtil.getInstanceId;
+import static org.springframework.data.mongodb.core.query.Criteria.where;
+import static org.springframework.data.mongodb.core.query.Query.query;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.bson.Document;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.client.model.IndexOptions;
+
+public class PersistedInstanceIdRecalculationService implements PersistedInstanceIdRecalculationServiceMBean {
+
+	private final Logger log = LoggerFactory.getLogger(getClass());
+	private final SpaceMirrorContext spaceMirror;
+
+	public PersistedInstanceIdRecalculationService(SpaceMirrorContext spaceMirror) {
+		this.spaceMirror = spaceMirror;
+	}
+
+	@Override
+	public void recalculatePersistedInstanceId(String collectionName, int numberOfInstances) {
```

Comment:
The idea was that if the number of instances change, the next load would not be able to use the index, but after loading we could run the `recalculatePersistedInstanceId` to update the data for subsequent loads. I was thinking this could be done while the PU was running.

Others have suggested that this could be done by touching all objects in the space, forcing them to be updated to mongo. The advantage of this is that it would require no special code for recalculating the instance id field, other than some generic code to touch all objects. This would also be done after deploying with a new instance count, where the initial load would have to be done without using this index.

So this whole PR only helps the load times when the instance count is unchanged. When it is changed the old `loadDocumentsRouted` could be used.